packages/client/src/objectSet/ObjectSetListenerWebsocket.ts

```diff
@@ -72,74 +105,107 @@ export class ObjectSetListenerWebsocket {
+    Subscription
+  >();
 
-  #conjureContext: ConjureContext;
+  #oswContext: ConjureContext;
   #metadataContext: ConjureContext;
   #ossContext: ConjureContext;
 
+  #maybeDisconnectTimeout: ReturnType<typeof setTimeout> | undefined;
+
   private constructor(client: MinimalClient) {
     this.#client = client;
+    this.#logger = client.logger?.child({}, {
+      msgPrefix: "<OSW> ",
+    });
+    invariant(
+      client.stack.startsWith("https://") || client.stack.startsWith("http://"),
+      "Stack must be a URL",
+    );
 
-    const stackUrl = new URL(client.stack);
-    this.#conjureContext = {
-      baseUrl: stackUrl.origin,
+    this.#oswContext = {
+      baseUrl: client.stack,
       servicePath: "/object-set-watcher/api",
       fetchFn: client.fetch,
       tokenProvider: async () => await client.tokenProvider(),
     };
     this.#ossContext = {
-      baseUrl: stackUrl.origin,
+      ...this.#oswContext,
       servicePath: "/object-set-service/api",
-      fetchFn: client.fetch,
-      tokenProvider: async () => await client.tokenProvider(),
     };
     this.#metadataContext = {
-      baseUrl: stackUrl.origin,
+      ...this.#oswContext,
       servicePath: "/ontology-metadata/api",
-      fetchFn: client.fetch,
-      tokenProvider: async () => await client.tokenProvider(),
     };
   }
 
-  subscribe<Q extends ObjectOrInterfaceDefinition>(
+  async subscribe<Q extends ObjectOrInterfaceDefinition>(
     objectSet: ObjectSet,
     listener: ObjectSetListener<Q>,
-  ): () => void {
-    const requestId = crypto.randomUUID();
-    const expiry = setTimeout(() => {
-      this.#expire(requestId);
-    }, ONE_DAY_MS);
-    this.#listeners.set(requestId, { listener, objectSet, expiry });
-    this.#subscribe(requestId, objectSet);
+  ): Promise<() => void> {
+    globalThis.crypto ??= (await import("node:crypto")).webcrypto as any;
+    const sub: Subscription = {
+      listener: fillOutListener<any>(listener),
+      objectSet,
+      status: "preparing",
+
+      // Since we don't have a real subscription id yet but we need to keep
+      // track of this reference, we can just use a random uuid.
+      subscriptionId: `TMP-${crypto.randomUUID()}`,
+    };
+    this.#subscriptions.set(sub.subscriptionId, sub);
+
+    // actually prepares the subscription, ensures the ws is ready, and sends
+    // a subscribe message. We don't want to block on this.
+    this.#initiateSubscribe(sub);
+
     return () => {
-      this.#unsubscribe(requestId);
+      this.#unsubscribe(sub);
     };
   }
 
-  async #subscribe(requestId: string, objectSet: ObjectSet) {
+  /**
+   * Called at least once for every subscription.
+   *
+   * - Resets pending expiry
+   * - Recreates temporary object set
+   * - Triggers a full subscribe message
+   *
+   * @returns
+   */
+  async #initiateSubscribe(sub: Subscription) {
+    if (sub.expiry) {
+      clearTimeout(sub.expiry);
+    }
+    sub.expiry = setTimeout(() => this.#expire(sub), ONE_DAY_MS);
```

Comment:
should the expiration time be configurable? (with this as a fallback)

Reply:
The temporary object set we create is set to 1 day. In reality this would be hidden by gateway but we have to deal in these directly for now.

Reply:
ah ok - comment? :)

Reply:
Will be in next update